polylang.php

```diff
@@ -36,25 +36,41 @@
  * along with this program. If not, see <https://www.gnu.org/licenses/>.
  */
 
-if ( ! defined( 'ABSPATH' ) ) {
-	exit; // Don't access directly.
-}
+defined( 'ABSPATH' ) || exit;
 
 if ( defined( 'POLYLANG_VERSION' ) ) {
 	// The user is attempting to activate a second plugin instance, typically Polylang and Polylang Pro.
 	require_once ABSPATH . 'wp-admin/includes/plugin.php';
 	require_once ABSPATH . 'wp-includes/pluggable.php';
+
 	if ( is_plugin_active( plugin_basename( __FILE__ ) ) ) {
 		deactivate_plugins( plugin_basename( __FILE__ ) ); // Deactivate this plugin.
 		// WP does not allow us to send a custom meaningful message, so just tell the plugin has been deactivated.
 		wp_safe_redirect( add_query_arg( 'deactivate', 'true', remove_query_arg( 'activate' ) ) );
 		exit;
 	}
-} else {
-	// Go on loading the plugin.
-	define( 'POLYLANG_VERSION', '3.8-dev' );
-	define( 'PLL_MIN_WP_VERSION', '6.2' );
-	define( 'PLL_MIN_PHP_VERSION', '7.2' );
+	return;
+}
+
+// Stopping here if we are going to deactivate the plugin (avoids breaking rewrite rules).
+if ( ! empty( $_GET['deactivate-polylang'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification
+	return;
+}
+
+require_once __DIR__ . '/install/install-base.php';
+require_once __DIR__ . '/install/install.php';
+$install = new PLL_Install(
+	__FILE__, // Plugin file.
+	'3.8-dev', // Plugin version.
+	'6.2', // WP version.
+	'7.2' // PHP version.
+);
```

Comment:
I'm not fan to replace the constants by parameters in this case. I often change their values and it's easier for me as this was earlier. Moreover I'm not comfortable to change the behavior of these constant definitions (they are defined this way for a very long time. Who knows how they are used?)

Although I don't see new tests, I guess your goal is to make the more code in the class `PLL_Install` testable. I suggest to read constants with `pll_get_constant()` to be able to mock constants instead.

Reply:
Constants are back in [5960897](https://github.com/polylang/polylang/pull/1704/commits/59608970dff0f8b5910ae090a265ced0b9884670).